demo-circuit/src/naive_threshold_sig_w_key_rotation/tests/next_cert.rs

```diff
@@ -252,7 +282,125 @@ fn master_key_rotation_works() {
+    const ITER: usize = 5;
+
+    let mut rng = thread_rng();
+    let ck_g1 = init_g1_committer_key();
+    let setup_circuit = NaiveThresholdSignatureWKeyRotation::get_instance_for_setup(MAX_PKS, 1);
+    let params = CoboundaryMarlin::index(&ck_g1, setup_circuit).unwrap();
+    let (
+        mut signing_keys_sks,
+        master_keys_sks,
+        mut genesis_validator_keys_tree_root,
+        mut withdrawal_certificate,
+        mut wcert_signatures,
+        mut validator_key_updates,
+    ) = setup_certificate_data(MAX_PKS, ITER);
+
+    let mut updated_signing_keys_sks = signing_keys_sks.clone();
+    let mut prev_withdrawal_certificate = create_withdrawal_certificate();
+    prev_withdrawal_certificate.custom_fields[0] = genesis_validator_keys_tree_root;
+    prev_withdrawal_certificate.quality = ITER as u64;
+    let threshold = ITER;
+
+    for i in 1..=ITER {
+        println!("iteration {}", i);
+        let circuit_res = NaiveThresholdSignatureWKeyRotation::new(
+            validator_key_updates.clone(),
+            wcert_signatures.clone(),
+            withdrawal_certificate.clone(),
+            Some(prev_withdrawal_certificate.clone()),
+            threshold as u64,
+            genesis_validator_keys_tree_root,
+        );
+        assert!(circuit_res.is_ok());
+        let circuit = circuit_res.unwrap();
+
+        let debug_res = debug_circuit(circuit.clone());
+        assert!(debug_res.is_ok());
+        let failing_constraint = debug_res.unwrap();
+        assert_eq!(failing_constraint, None);
+
+        let proof =
+            CoboundaryMarlin::prove(&params.0.clone(), &ck_g1, circuit.clone(), false, None)
+                .unwrap();
+
+        let public_inputs = [
+            circuit.genesis_constant,
+            get_cert_data_hash(&withdrawal_certificate),
+            get_cert_data_hash(&prev_withdrawal_certificate),
+        ];
+
+        assert!(
+            CoboundaryMarlin::verify(&params.1.clone(), &ck_g1, &public_inputs, &proof).unwrap()
+        );
+        signing_keys_sks = updated_signing_keys_sks.clone();
+        validator_key_updates.signing_keys = validator_key_updates.updated_signing_keys.clone();
+
+        prev_withdrawal_certificate = withdrawal_certificate.clone();
+        validator_key_updates
+            .updated_signing_keys_sk_signatures
+            .iter_mut()
+            .for_each(|s| *s = NULL_CONST.null_sig);
+        validator_key_updates
+            .updated_signing_keys_mk_signatures
+            .iter_mut()
+            .for_each(|s| *s = NULL_CONST.null_sig);
+        validator_key_updates
+            .updated_master_keys_sk_signatures
+            .iter_mut()
+            .for_each(|s| *s = NULL_CONST.null_sig);
+        validator_key_updates
+            .updated_master_keys_mk_signatures
+            .iter_mut()
+            .for_each(|s| *s = NULL_CONST.null_sig);
+
+        let sig_key_changes = rng.gen::<usize>() % ITER + 1;
```

Comment:
Why never changing all the signing keys? Maybe a different quality may be employed in different iterations so that all the signing keys are used and so it becomes meaningful to rotate all of them

Reply:
Done

Reply:
I meant to modify this test to have a more dynamic example where different number of validators may sign the certificate in each round. So you could sample `sig_key_changes` in `1..MAX_PKS` and sample in each round the quality of the certificate in the range `threshold..MAX_PKS`